src/dwfl_thread_callbacks.cc

```diff
@@ -13,16 +13,80 @@ pid_t next_thread(Dwfl *dwfl, void *arg, void **thread_argp) {
   return us->pid;
 }
 
+// DWARF and the Linux kernel don't have a uniform view of the processor, so we
+// can't just follow the register mask and shove it into the output registers.
+// Instead, we crib off of libdwfl's ARM/x86 unwind code in elfutil's
+// libdwfl/unwind-libdw.c
 bool set_initial_registers(Dwfl_Thread *thread, void *arg) {
+  Dwarf_Word regs[33] = {}; // max register count across all arcs
+  uint64_t n = 0;
   struct UnwindState *us = reinterpret_cast<UnwindState *>(arg);
-  Dwarf_Word regs[17] = {0};
 
-  // Only 3 registers are used in the unwinding
-  regs[6] = us->initial_regs.ebp;
-  regs[7] = us->initial_regs.esp;
-  regs[16] = us->initial_regs.eip;
+#ifdef __x86_64__
+  // Substantial difference here in 32- and 64-bit x86; only support 64-bit now
+  regs[n++] = us->initial_regs.regs[REGNAME(EAX)];
+  regs[n++] = us->initial_regs.regs[REGNAME(EDX)];
+  regs[n++] = us->initial_regs.regs[REGNAME(ECX)];
+  regs[n++] = us->initial_regs.regs[REGNAME(EBX)];
+  regs[n++] = us->initial_regs.regs[REGNAME(ESI)];
+  regs[n++] = us->initial_regs.regs[REGNAME(EDI)];
+  regs[n++] = us->initial_regs.regs[REGNAME(EBP)];
+  regs[n++] = us->initial_regs.regs[REGNAME(ESP)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R8)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R9)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R10)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R11)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R12)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R13)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R14)];
+  regs[n++] = us->initial_regs.regs[REGNAME(R15)];
```

Comment:
I think x86 will not work. we are miss-aligned here. initial regs are not as big as the reg structure.
The naming is nice to document and debug things, though having a generic copy structure might be easier. I will let you decide what you prefer.

Reply:
I don't see any changes here.  Is the repo missing a commit?

Reply:
Changes are actually pushed to the unwind_state structure to align things.